### PR TITLE
home-manager: use coreutils for readlink

### PR DESCRIPTION
**Pull request by andreykaipov**
### Description

I'm using HM on macOS and ran into some issues:

```
Activating setupLaunchAgents
readlink: illegal option -- m
usage: readlink [-fn] [file ...]
readlink: illegal option -- m
usage: readlink [-fn] [file ...]
```

Similar issue to https://github.com/nix-community/home-manager/issues/3516 which was fixed in https://github.com/nix-community/home-manager/pull/3517, but there were a few more readlink commands in the existing Nix code. 

Also related: https://github.com/nix-community/home-manager/pull/11

There might be more tools conflict on macOS that need to be replaced with GNU equivalents from coreutils, but readlink was what I ran into.

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.

  - I seem to get the same build failure when running the tests against my branch and against master, so I'd say they pass. 😅 

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://nix-community.github.io/home-manager/#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

#### Maintainer CC

@rycee

### CODE REVIEW

**Comment by rycee** [comment]
Thanks for the report. Would you mind trying to add

    home.emptyActivationPath = true;

to your configuration and then do a HM switch without this patch?

Also, what is the value of your `home.stateVersion` option?

**Comment by andreykaipov** [comment]
Hello! Sorry for the slow response!

So with `home.emptyActivationPath = true;` explicitly set I still got the same `readlink: illegal option -- m` error.

My `home.stateVersion` is 22.11 and I saw here that means it should be true for me: https://github.com/nix-community/home-manager/blob/601c22f8af61b9a2e47001474c595c78771aa3b2/modules/home-environment.nix#L354-L360

So I tried setting it to false, but I strangely got the same issue.

I then added `echo "$PATH"` to one one of activation scripts to compare the outputs of each. In both cases I saw `/bin:/usr/bin:~/.nix-profile/bin` prepended to my path and was puzzled for a while until I remembered that 8 months ago I tried to get fancy with my home manager config and abstracted my activation script entrypoint:

https://github.com/andreykaipov/self/commit/a9eccd78db4dac6f3899643ce0a524ad08711d5c#diff-70bea2658bcd8c1952d4a1c3590bcc07fde1755308e680317c2559432bd7f4d8R36-R46

Where I literally **prepend** exactly those paths. And I distinctly remember the reason I did that at the time was because most of the binaries in my activation scripts weren't recognized otherwise! So in hindsight all I really needed to do was set `home.emptyActivationPath = false;`! Or just append my paths instead of prepend... And I also see now there's a `extraActivationPath` variable that can do the same thing.

In any case, this all happened on a Linux box, so I guess I never noticed any quirks there, but only when I tried using the same configs on this macOS machine.

---

So TLDR is I'm an idiot. However maybe there is still merit in making these readlink calls use coreutils so that it can still work regardless of how much a user might mangle their path? 😅

**Comment by rycee** [comment]
Ok, glad you found the source of these calls to readline. I'm quite reluctant to change the existing behavior, though. Especially for coreutils, which are used _everywhere_.

One possibility would be to mark the `PATH` variable readonly, which would prevent people from messing around with it, instead they would get an error message `PATH: readonly variable`. I'll have to think a bit about it.

In any case, I think I'll close this PR, thanks for contributing, though. And thanks for the clear explanation about what happened!

**Comment by MXfive** [comment]
Hitting this issue also, but without doing anything myself to the PATH in the activation script.  What's strange is `PATH` looks correct when I add it into the activation script.

Modified activate script
```bash
  echo $PATH
  echo "$(which readlink)"
  newDir="$(readlink -m "$newGenPath/LaunchAgents")"
```

```
Activating setupLaunchAgents
/nix/store/d72z8b7fvvcin2p84x7h5s0abkggplzc-coreutils-9.7/bin:OTHER_NIX_PATHS:/opt/homebrew/bin:/opt/homebrew/sbin:/Users/martyn/.local/bin:/Users/martyn/.krew/bin:/Users/martyn/go/bin:/Users/martyn/.nix-profile/bin:/run/current-system/sw/bin:/nix/var/nix/profiles/default/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
/usr/bin/readlink
readlink: illegal option -- m
usage: readlink [-fn] [file ...]
readlink: illegal option -- m
usage: readlink [-fn] [file ...]
find: : No such file or directory
```

Replacing with /nix/store/d72z8b7fvvcin2p84x7h5s0abkggplzc-coreutils-9.7/bin/readlink works without issue.


@rycee I think it would be more robust to explicity call the correct `readlink` binary explicitly rather than implicitly rely on $PATH being exactly how it needs to be.

**Comment by MXfive** [comment]
Update, found the culprit in my case, was part of my redacted PATH, Krew does this:

```
315:export PATH="$HOME/.krew/bin:/nix/store/9pa6rygnkrdp03j4ba0w6yb5qid5x3qb-git-2.49.0/bin:/usr/bin:$PATH";
```

So really do think it's worth having this be explicity in home-manager to not be fragile to badly written neighbour dependencies users may pull in.

Edit: in this case it was actually being pulled in by my config, Krew was part of another nix dotfiles that I'd taken some program configs from.

Still would be good to have home-manager be more robust and explicit IMO though.
